File: src/Python/247.StrobogrammaticNumberII.py

class Solution:
    def buildResult(self, curPos, n, cache, selfarr, diffMap, result):
        if curPos < (n + 1) // 2:
            for c in selfarr:
                if curPos == 0 and c == '0' and n != 1:
                    continue
                cache[curPos] = cache[n - 1 - curPos] = c
                self.buildResult(curPos + 1, n, cache, selfarr, diffMap, result)
            if n % 2 == 0 or curPos != n / 2:
                for c in diffMap.keys():
                    cache[curPos] = c
                    cache[n - 1 - curPos] = diffMap[c]
                    self.buildResult(curPos + 1, n, cache, selfarr, diffMap, result)
            elif n != 0:
                result.append(''.join(cache))


    def findStrobogrammatic(self, n):
        diffMap = {}
        diffMap['6'] = '9'
        diffMap['9'] = '6'
        selfarr = ['0', '1', '8']
        result = []
        cache = ['0'] * n
        self.buildResult(0, n, cache, selfarr, diffMap, result)

    
if __name__ == "__main__":
    solution = Solution()

File: src/CSharp/247.StrobogrammaticNumberII.cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.StrobogrammaticNumberII
{

public class Solution {
    public List<string> FindStrobogrammatic(int n) {
        var diffMap = new Dictionary<char, char>();
        char[] selfArr = new char[] { '0', '1', '8' };
        diffMap.Add('6', '9');
        diffMap.Add('9', '6');
        List<string> result = new List<string>();
        buildResult(0, n, new char[n], selfArr, diffMap, result);
        return result;
    }

    private void buildResult(int curPos, int n, char[] cache, char[] selfArr, Dictionary<char, char> diffMap, List<string> result) {
        if (curPos < (n + 1) / 2) {
            foreach (char c in selfArr) {
                if (curPos == 0 && c == '0' && n != 1)
                    continue;
                cache[curPos] = cache[n - 1 - curPos] = c;
                buildResult(curPos + 1, n, cache, selfArr, diffMap, result);
            }
            if (n % 2 == 0 || curPos != n / 2) {
                foreach (char c in diffMap.Keys) {
                    cache[curPos] = c;
                    cache[n - 1 - curPos] = diffMap[c];
                    buildResult(curPos + 1, n, cache, selfArr, diffMap, result);
                }
            }
        } else if (n != 0) {
            result.Add(new string(cache));
        }
    }
} 

}